include/revng/ADT/Concepts.h

```diff
@@ -4,44 +4,178 @@
+}
+
+} // namespace revng::detail
+
+// clang-format off
+template<class T>
+concept TupleLike = (TupleSizeCompatible<T>
+                     and revng::detail::checkAllTupleElementTypes<T>());
+// clang-format on
+
+static_assert(TupleLike<std::tuple<>>);
+static_assert(TupleLike<std::tuple<int, int, long>>);
+static_assert(TupleLike<std::pair<int, int>>);
+static_assert(TupleLike<std::array<int, 0>>);
+static_assert(not TupleLike<std::vector<int>>);
+
+//
+// Concepts to simplify working with specializations of templates.
+//
+
+/// A concept that helps determine whether a given object is a specialization
+/// (or is inheriting a specialization) of a given template.
+///
+/// This lets us cut down on the number of `IsX` (e.g. `IsGenericGraph`,
+/// `IsRank`) concepts needed since we can now just say
+/// `SpecializationOf<GenericGraph>` on the interface boundaries and get
+/// the expected check.
+///
+/// TODO: this requires clang-13+, uncomment it after the update.
+// template<typename Type, template<typename...> class Ref>
+// concept SpecializationOf = requires(Type &&Value) {
+//   []<typename... Ts>(Ref<Ts...> &){}(Value);
+// };
+//
+// template<typename Type, template<typename...> class Ref>
+// concept NotSpecializationOf = !SpecializationOf<Type, Ref>;
+//
+// static_assert(SpecializationOf<std::pair<int, long>, std::pair>);
+// static_assert(SpecializationOf<const std::pair<int, long>, std::pair>);
+// static_assert(SpecializationOf<std::string, std::basic_string>);
+// static_assert(SpecializationOf<std::ostream, std::basic_ios>);
+// static_assert(not SpecializationOf<std::string, std::basic_string_view>);
+
+/// TODO: Remove after updating to clang-format with concept support.
+struct ClangFormatPleaseDoNotBreakMyCode;
+// clang-format off
+// clang-format on
+
+namespace revng::detail {
+
+template<typename Test, template<typename...> class Ref>
+struct StrictSpecializationHelper : std::false_type {};
+
+template<template<typename...> class Ref, typename... Args>
+struct StrictSpecializationHelper<Ref<Args...>, Ref> : std::true_type {};
+
+template<template<typename...> class Ref, typename... Args>
+struct StrictSpecializationHelper<const Ref<Args...>, Ref> : std::true_type {};
+
+template<typename Test, template<typename...> class Ref>
+constexpr bool
+  StrictSpecialization = StrictSpecializationHelper<Test, Ref>::value;
+
+} // namespace revng::detail
+
+/// A more strict version of the `SpecializationOf` concept. This one only
+/// allows direct specializations (and aliases) - no inheritance.
+///
+/// It's useful in the cases when template parameter deduction is important,
+/// e.g. when instanciating traits.
+template<typename Test, template<typename...> class Ref>
+concept StrictSpecializationOf = revng::detail::StrictSpecialization<Test, Ref>;
+
+template<typename Test, template<typename...> class Ref>
+concept NotStrictSpecializationOf = !StrictSpecializationOf<Test, Ref>;
```

Comment:
This kind of concepts are misleading. Does the `Not` applies to `Strict` or to the whole concept?
Consider:
* Using `requires` locally with `!`
* `Not` -> `Negate`
* Pushing the concept close to its usage so that is immediately evident what's going on

Reply:
Actually, I prefer `requires` with `!` hands down. The only reason I even went with this temporarily is because all of those `clang-format off` are starting to get out of hand and I'm afraid they might be causing some problems of their own.

Reply:
ACK, still.

Reply:
Do you want to use this chance and pull the trigger on getting rid of all the `NotXXX` concepts then? I considered doing it, but didn't go through with it.

Reply:
For reference, this is what it would look like: https://github.com/revng/revng/pull/197/commits/e2d9a9bcf7eadcc49dfa5f5b018ce7077a1ed82d

Should I squash this commit? Or drop it?

Reply:
LGTM

Reply:
Squashed.